File: Java/Array/Lab Test 01 Practice/Solve/StudentRankingSystem/src/StudentManager.java
public class StudentManager {

    private Student[] studentList;
    private int studentCount;

    public StudentManager(){
        studentList = new Student[5];
        studentCount = 0;
    }

    public void insertStudent(Student newStudent) {
        if (studentCount < studentList.length || newStudent.getGpa() > studentList[studentList.length - 1].getGpa()) {
            if (studentCount == studentList.length) {
                deleteStudent(studentList[studentList.length - 1].getId()); // Remove the student with the lowest GPA
            }

            int i;
            for (i = studentCount - 1; i >= 0; i--) {
                if (studentList[i] != null && (studentList[i].getGpa() < newStudent.getGpa() ||
                        (studentList[i].getGpa() == newStudent.getGpa() && studentList[i].getId() > newStudent.getId()))) {
                    studentList[i + 1] = studentList[i];
                } else {
                    break;
                }
            }
            studentList[i + 1] = newStudent;
            studentCount++;
        }
    }

    public void deleteStudent(int id) {
        int i;
        for (i = 0; i < studentCount; i++) {
            if (studentList[i].getId() == id) {
                break;
            }
        }
        if (i < studentCount) {
            for (int j = i; j < studentCount - 1; j++) {
                studentList[j] = studentList[j + 1];
            }
            studentList[studentCount - 1] = null;
            studentCount--;
        } else {
            System.out.println("Student not found.");
        }
    }

    public void updateStudent(int id, String newName, double newGpa) {
        deleteStudent(id);
        insertStudent(new Student(id, newName, newGpa));
    }

    public Student[] getTopStudents() {
        Student[] topStudents = new Student[Math.min(studentCount, 5)];
        for (int i = 0; i < topStudents.length; i++) {
            topStudents[i] = studentList[i];
        }
        return topStudents;
    }
}

File: Python/Array/Lab 01 Practice/StudentRankingSystem/StudentManager.py
from StudentRankingSystem.Student import Student

class StudentManager:
    def __init__(self):
        self.studentList = [None] * 5
        self.studentCount = 0

    def insert_student(self, newStudent):
        if self.studentCount < len(self.studentList) or newStudent.getGpa() > self.studentList[self.studentCount-1].getGpa():
            if self.studentCount == len(self.studentList):
                self.delete_student(self.studentList[len(self.studentList)-1].getId())

            i = self.studentCount - 1
            while i >= 0 and (self.studentList[i] is not None and (self.studentList[i].getGpa() < newStudent.getGpa() or (self.studentList[i].getGpa() == newStudent.getGpa() and self.studentList[i].getId() > newStudent.getId()))):
                if i + 1 < len(self.studentList):
                    self.studentList[i + 1] = self.studentList[i]
                i -= 1
            
            if i + 1 < len(self.studentList):
                self.studentList[i + 1] = newStudent
            if self.studentCount < len(self.studentList):
                self.studentCount += 1

    def delete_student(self, id):
        for i in range(len(self.studentList)):
            if self.studentList[i] is not None and self.studentList[i].getId() == id:
                for j in range(i, len(self.studentList) - 1):
                    self.studentList[j] = self.studentList[j + 1]
                self.studentList[len(self.studentList) - 1] = None
                self.studentCount -= 1
                break

    def update_student(self, newId, newName, newGpa):
        self.delete_student(newId)
        newStudent = Student(newId, newName, newGpa)
        self.insert_student(newStudent)

    def get_top_students(self):
        topStudentList = []
        for i in range(min(self.studentCount, 5)):
            topStudentList.append(self.studentList[i])
        return topStudentList